src/Batch.php

```diff
@@ -36,8 +37,14 @@ public function dispatch(object $message, array $stamps = []): Envelope
 
         $envelope = $this->wrappedBus->dispatch($envelope);
 
-        if (($stamp = $envelope->last(Deferred::class)) !== null) {
-            $this->transportsToFlush[] = $stamp->getTransport();
+        if (($sentStamp = $envelope->last(SentStamp::class)) !== null) {
+            if (($deferredStamp = $envelope->last(Deferred::class)) !== null) {
+                if (($senderAlias = $sentStamp->getSenderAlias()) !== null) {
+                    $this->transportsToFlush[$senderAlias] = $deferredStamp->getTransport();
+                } else {
+                    $this->transportsToFlush[] = $deferredStamp->getTransport();
```

Comment:
it seems the `senderAlias` can actually be NULL.

Reply:
Oh I forgot about this. Nicolas did mention about doing deduplication. Since these are services, what do you think about deduplicating with the spl_object_hash or we can do the dedup with an SplObjectMap instead of an array.

Reply:
updated to use `spl_object_id`